_posts/2022-09-15

```diff
@@ -0,0 +1,128 @@
+---
+layout: post
+title: Disposables and Dependency Injection
+---
+
+Chances that you've been using a dependency injection container in a C# application are extremely high. All DI containers I'm aware of provide the capability of a _child container_ or _scope_ as the Microsoft DI abstraction calls it. Scopes are providing a dedicated instance lifetime controlled by the user. Consider the following code:
+
+```csharp
+interface I1
+{
+}
+
+interface I2 : I1
+{
+}
+
+class Impl : I2, IDisposable
+{
+    public static int DisposeCounter { get; set; }
+
+    public void Dispose()
+    {
+        DisposeCounter++;
+    }
+}
+```
+
+```csharp
+serviceCollection.AddScoped<Impl>();
+
+// build IServiceProvider
+
+using (var scope = serviceProvider.CreateScope())
+{
+    var service = scope.ServiceProvider.GetService<Impl>();
+}
+```
+
+Disposing the scope will also call dispose on our instance of `MyService` since it implements `IDisposable`. That's not too surprising, but let's look at some more complicated scenarios:
+
+What will `DisposeCounter` be if the resolved type does not implement `IDisposable` (but the implementation does)?
+
+```csharp
+serviceCollection.AddScoped<I1>(_ => new Impl());
+
+// build IServiceProvider
+
+using (var scope = serviceProvider.CreateScope())
+{
+    var service = scope.ServiceProvider.GetService<I1>();
+}
+```
+
+And does `DisposeCounter` change if we resolve the instance from the container too?
+
+```csharp
+serviceCollection.AddScoped<I1>(sp => sp.GetRequiredService<Impl>());
+serviceCollection.AddScoped<Impl>();
+
+// build IServiceProvider
+
+using (var scope = serviceProvider.CreateScope())
+{
+    var service = scope.ServiceProvider.GetService<I1>();
+}
+```
+
+And what happens if we resolve another interface which provides no compile-time information about the resolved implementation?
+
+```csharp
+serviceCollection.AddScoped<I1>(sp => sp.GetRequiredService<I2>());
+serviceCollection.AddScoped<I2>(sp => sp.GetRequiredService<Impl>());
+serviceCollection.AddScoped<Impl>();
+
+// build IServiceProvider
+
+using (var scope = serviceProvider.CreateScope())
+{
+    var service = scope.ServiceProvider.GetService<I1>();
+}
+```
+
+As always, the answer is "it depends". In our case, it depends on the dependency injection container used.
+
+## Microsoft ServiceProvider
+
+Let's start with the Microsoft ServiceProvider. By now it's not just one of the most used DI containers but also pretty much acts as a reference implementation for many other DI containers.
+
+If we build the `IServiceProvider` using `var serviceProvider = serviceCollection.BuildServiceProvider()` we get the following `DisposeCounter` values:
+
+1. 1
+1. 2
+1. 3
```

Comment:
Maybe add a comment about what scenarios 1, 2, and 3 were

Reply:
added some text, does it read better this way?